lib/matplotlib/_constrained_layout.py

```diff
@@ -83,181 +74,357 @@ def do_constrained_layout(fig, renderer, h_pad, w_pad,
+        for cbax in ax._colorbars:
+            # note pad is a fraction of the parent width...
+            pad = _colorbar_get_pad(cbax)
+            # colorbars can be child of more than one subplot spec:
+            cbp_rspan, cbp_cspan = _get_cb_parent_spans(cbax)
+            loc = cbax._colorbar_info['location']
+            if loc == 'right':
+                if cbp_cspan.stop == ss.colspan.stop:
+                    # only increase if the colorbar is on the right edge
+                    cbpos, cbbbox = _get_pos_and_bbox(cbax, renderer)
+                    margin['rightcb'] += cbbbox.width + pad
+            elif loc == 'left':
+                if cbp_cspan.start == ss.colspan.start:
+                    # only increase if the colorbar is on the left edge
+                    cbpos, cbbbox = _get_pos_and_bbox(cbax, renderer)
+                    margin['leftcb'] += cbbbox.width + pad
+            elif loc == 'top':
+                if cbp_rspan.start == ss.rowspan.start:
+                    cbpos, cbbbox = _get_pos_and_bbox(cbax, renderer)
+                    margin['topcb'] += cbbbox.height + pad
+            else:
+                if cbp_rspan.stop == ss.rowspan.stop:
+                    cbpos, cbbbox = _get_pos_and_bbox(cbax, renderer)
+                    margin['bottomcb'] += cbbbox.height + pad
+        # pass the new margins down to the layout grid for the solution...
+        gs._layoutgrid.edit_outer_margin_mins(margin, ss)
+
+
+def _make_margin_suptitles(fig, renderer, *, w_pad=0, h_pad=0):
+    # Figure out how large the suptitle is and make the
+    # top level figure margin larger.
+    for panel in fig.panels:
+        _make_margin_suptitles(panel, renderer, w_pad=w_pad, h_pad=h_pad)
+    invTransFig = fig.transPanel.inverted().transform_bbox
+    pan2fig = fig.transPanel + fig.transFigure.inverted()
+
+    w_pad, h_pad = pan2fig.transform((w_pad, h_pad))
+
+    if fig._suptitle is not None and fig._suptitle.get_in_layout():
+        bbox = invTransFig(fig._suptitle.get_tightbbox(renderer))
+        p = fig._suptitle.get_position()
+        fig._suptitle.set_position((p[0], 1-h_pad))
+        fig._layoutgrid.edit_margin_min('top', bbox.height + 2 * h_pad)
+
+    if 0:
+        # Not implimented yet!
+        if fig._supxlabel is not None:
+            bbox = invTransFig(fig._supxlabel.get_tightbbox(renderer))
+            fig._layoutgrid.edit_margin_min('bottom', bbox.height + 2 * h_pad)
+
+        if fig._supylabel is not None:
+            bbox = invTransFig(fig._supylabel.get_tightbbox(renderer))
+            fig._layoutgrid.edit_margin_min('left', bbox.width + 2 * w_pad)
+
+
+def _match_submerged_margins(fig):
+    """
+    Make the margins that are submerged inside an axes the same
+    size.
+
+    This allows axes that span two columns (or rows) that are offset
+    from one another to have the same size.
+
+    i.e. if in row 0, the axes is at columns 0, 1 and for row 1,
```

Comment:
asciiart may help?

Reply:
Improved the doc of this because even I forgot what it was doing exactly...